tls/s2n_handshake_io.c

```diff
@@ -441,6 +444,52 @@ int s2n_conn_set_handshake_no_client_cert(struct s2n_connection *conn) {
     return 0;
 }
 
+const char *s2n_connection_get_handshake_type_name(struct s2n_connection *conn) 
+{
+    notnull_check_ptr(conn);
+
+    int handshake_type = conn->handshake.handshake_type;
+
+    if (handshake_type_str[handshake_type][0] != '\0') {
+        return handshake_type_str[handshake_type];
+    }
+
+    /* Compute handshake_type_str[handshake_type] */
+    char *p = handshake_type_str[handshake_type];
+    char *end = p + sizeof(handshake_type_str[0]);
+    if (handshake_type > INITIAL) {
```

Comment:
instead of a list of if statements, have you considered keeping both the handshake types and string representations in an array, then calculate the string representation by iterating through the array and comparing bitwise?

eg. 
```c
int handshake_types[] = { INITIAL, NEGOTIATED, FULL_HANDSHAKE...}
char handshake_type_names[][] = { "INITIAL", "NEGOTIATED", "FULL_HANDSHAKE"...}
```
or 
```c
handshake_type_names[][] = { [INITIAL] = "INITIAL", [NEGOTIATED] = "NEGOTIATED"... }
```
then
```c
for (int i = 0; i < size; i++) {
   if (handshake_type >> (i - 1) & 1) ...
}
```

Reply:
> instead of a list of if statements, have you considered keeping both the handshake types and string representations in an array, then calculate the string representation by iterating through the array and comparing bitwise?
I like your idea to eliminate if statements. I did not feel very strong about this though. 

There are pro and cons:
My approach:
Pros:
Simple and faster. Space complexity for those ifs O(1). 
Cons:
Maybe we think it will be too many ifs but how many bits will be there in future. Currently, only 7 bits up to 0x40. I think 7 ifs is fine.  

The proposed approach:
Pros: 
Code could be shorter but the code example is wrong...
Cons:
1st example: two array, space complexity O(n).
2nd example: one array using 0x40 as an array index... space complexity O(n * n)
What do we gain here if we do not have cache?

> 
> eg.
> 
> ```c
> int handshake_types[] = { INITIAL, NEGOTIATED, FULL_HANDSHAKE...}
> char* handshake_type_names[] = { "INITIAL", "NEGOTIATED", "FULL_HANDSHAKE"...}
> ```
> 
in c, we use const char* to represent const string
```c
const char* handshake_type_names[] = { "INITIAL", "NEGOTIATED", "FULL_HANDSHAKE"...}
```

>
> or
> 
> ```c
> handshake_type_names[][] = { [INITIAL] = "INITIAL", [NEGOTIATED] = "NEGOTIATED"... }
> ```
> 
We might want a good reason for wasted spaces.

> then
> 
> ```c
> for (int i = 0; i < size; i++) {
>    if (handshake_type >> (i - 1) & 1) ...
> }
> ```
This won't work as right shift 0 >> -1 & 1 will not compile... Plus, you still need to handle INITIAL(0x0) separately.